src/tests/crate_policies.rs

```diff
@@ -0,0 +1,145 @@
+        PackagePolicyEntry::Unversioned(Default::default()),
+    );
+    let store = Store::mock(config, audits, imports);
+    let cfg = mock_cfg(&metadata);
+
+    let result = crate::check_crate_policies(&cfg, &store);
+    assert_eq!(
+        result,
+        Err(CratePolicyErrors {
+            errors: vec![CratePolicyError::NeedsVersion(NeedsPolicyVersionErrors {
+                errors: vec![
+                    PackageError {
+                        package: "third-party".into(),
+                        version: Some(ver(1))
+                    },
+                    PackageError {
+                        package: "third-party".into(),
+                        version: Some(ver(2))
+                    }
+                ]
+            })]
+        })
+    );
+}
+
+/// Checks that if a third-party crate is present and a versioned policy is used for one version,
+/// an error occurs indicating that versions are needed for other versions (regardless of whether
+/// they are considered first- or third-party).
+#[test]
+fn simple_crate_policies_third_party_crates_need_all_versions() {
+    let _enter = TEST_RUNTIME.enter();
+
+    let mock = MockMetadata::overlapping();
+    let metadata = mock.metadata();
+
+    for (a, b) in [(1, 2), (2, 1)] {
+        let (mut config, audits, imports) = builtin_files_full_audited(&metadata);
+        config.policy.insert(
+            "third-party".into(),
+            PackagePolicyEntry::Versioned {
+                version: [(ver(a), Default::default())].into(),
+            },
+        );
+        let store = Store::mock(config, audits, imports);
+        let cfg = mock_cfg(&metadata);
+
+        let result = crate::check_crate_policies(&cfg, &store);
+        assert_eq!(
+            result,
+            Err(CratePolicyErrors {
+                errors: vec![CratePolicyError::NeedsVersion(NeedsPolicyVersionErrors {
+                    errors: vec![PackageError {
+                        package: "third-party".into(),
+                        version: Some(ver(b))
+                    }]
+                })]
+            })
+        );
+    }
+}
+
+/// If crate policies are provided for versions which aren't present in the graph, an error should
+/// occur.
+#[test]
+fn simple_crate_policies_extraneous_crate_versions() {
+    let _enter = TEST_RUNTIME.enter();
+
+    let mock = MockMetadata::overlapping();
+    let metadata = mock.metadata();
+    let (mut config, audits, imports) = builtin_files_full_audited(&metadata);
+    config.policy.insert(
+        "third-party".into(),
+        PackagePolicyEntry::Versioned {
+            version: [
+                (ver(1), Default::default()),
+                (ver(2), Default::default()),
+                (ver(3), Default::default()),
+            ]
+            .into(),
+        },
+    );
+    let store = Store::mock(config, audits, imports);
+    let cfg = mock_cfg(&metadata);
+
+    let result = crate::check_crate_policies(&cfg, &store);
```

Comment:
Historically we've traced this sort of error by doing a snapshot with `insta` like what we do for `store_parsing` (https://github.com/mozilla/cargo-vet/blob/61b7539f9c8b452597c307e632d976d869ef2682/src/tests/store_parsing.rs#L15-L27), that way we can see what the actual formatted errors look like easily in snapshots as well as checking the specific errors. 

I'm not necessarily opposed to doing explicit checking here, but it might be nice to also capture the errors with an insta snapshot so it's easier to see what they will look like.

Reply:
Okay. I was going to use insta, but this approach seemed more precise (as far as avoiding arbitrary string/output changes causing tests to fail). But OTOH keeping those similar or acknowledging when they change is useful too.

Reply:
FWIW it's very easy to re-generate insta snapshots, which is one of the nice things about that approach :-)

I'd also be fine with doing both this check and an insta snapshot as well if you'd prefer that.